File: Raven.Client.Java/src/main/java/net/ravendb/client/document/batches/LazyMultiLoaderWithInclude.java
package net.ravendb.client.document.batches;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

import net.ravendb.abstractions.basic.Lazy;
import net.ravendb.abstractions.basic.Tuple;
import net.ravendb.abstractions.closure.Function0;
import net.ravendb.abstractions.extensions.ExpressionExtensions;
import net.ravendb.client.IDocumentSessionImpl;

import com.google.common.base.Defaults;
import com.mysema.query.types.Path;


public class LazyMultiLoaderWithInclude implements ILazyLoaderWithInclude {
  private final IDocumentSessionImpl session;
  private final List<Tuple<String, Class<?>>> includes = new ArrayList<>();

  /**
   * Initializes a new instance of the {@link LazyMultiLoaderWithInclude} class.
   * @param session
   */
  public LazyMultiLoaderWithInclude(IDocumentSessionImpl session) {
    this.session = session;
  }

  /**
   * Includes the specified path.
   * @param path
   * @return
   */
  @Override
  public ILazyLoaderWithInclude include(String path) {
    includes.add(Tuple.<String, Class<?>> create(path, Object.class));
    return this;
  }

  /**
   * Includes the specified path.
   * @param path
   * @return
   */
  @Override
  public ILazyLoaderWithInclude include(Path<?> path) {
    return include(ExpressionExtensions.toPropertyPath(path));
  }

  @SuppressWarnings("unchecked")
  @Override
  public <TResult> Lazy<TResult[]> load(Class<TResult> clazz, String... ids) {
    return session.lazyLoadInternal(clazz, ids, includes.toArray(new Tuple[0]), null);
  }

  @SuppressWarnings("unchecked")
  @Override
  public <TResult> Lazy<TResult[]> load(Class<TResult> clazz, Collection<String> ids) {
    return session.lazyLoadInternal(clazz, ids.toArray(new String[0]), includes.toArray(new Tuple[0]), null);
  }

  @Override
  public <TResult> Lazy<TResult> load(final Class<TResult> clazz, String id) {
    final Lazy<TResult[]> lazy = load(clazz, Arrays.asList(id));
    return new Lazy<>(new Function0<TResult>() {
      @Override
      public TResult apply() {
        TResult[] results = lazy.getValue();
        return results.length > 0 ? results[0] : Defaults.defaultValue(clazz);
      }
    });
  }

  @Override
  public <TResult> Lazy<TResult> load(Class<TResult> clazz, Number id) {
    String documentKey = session.getConventions().getFindFullDocumentKeyFromNonStringIdentifier().find(id, clazz, false);
    return load(clazz, documentKey);
  }

  @Override
  public <TResult> Lazy<TResult> load(Class<TResult> clazz, UUID id) {
    String documentKey = session.getConventions().getFindFullDocumentKeyFromNonStringIdentifier().find(id, clazz, false);
    return load(clazz, documentKey);
  }

  @Override
  public <TResult> Lazy<TResult[]> load(Class<TResult> clazz, Number... ids) {
    List<String> documentKeys = new ArrayList<>();
    for (Number id: ids) {
      documentKeys.add(session.getConventions().getFindFullDocumentKeyFromNonStringIdentifier().find(id, clazz, false));
    }
    return load(clazz, documentKeys);
  }

  @Override
  public <TResult> Lazy<TResult[]> load(Class<TResult> clazz, UUID... ids) {
    List<String> documentKeys = new ArrayList<>();
    for (UUID id: ids) {
      documentKeys.add(session.getConventions().getFindFullDocumentKeyFromNonStringIdentifier().find(id, clazz, false));
    }
    return load(clazz, documentKeys);
  }




}

File: src/Raven.NewClient/Document/Batches/LazyMultiLoaderWithInclude.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using Raven.NewClient.Abstractions.Extensions;
using System.Threading.Tasks;

namespace Raven.NewClient.Client.Document.Batches
{
    public class LazyMultiLoaderWithInclude<T> : ILazyLoaderWithInclude<T>
    {
        private readonly IDocumentSessionImpl session;
        private readonly List<string> includes = new List<string>();

        /// <summary>
        /// Initializes a new instance of the <see cref="LazyMultiLoaderWithInclude{T}"/> class.
        /// </summary>
        /// <param name="session">The session.</param>
        internal LazyMultiLoaderWithInclude(IDocumentSessionImpl session)
        {
            this.session = session;
        }

        /// <summary>
        /// Includes the specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        public ILazyLoaderWithInclude<T> Include(string path)
        {
            includes.Add(path);
            return this;
        }

        /// <summary>
        /// Includes the specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        public ILazyLoaderWithInclude<T> Include(Expression<Func<T, string>> path)
        {
            return Include(path.ToPropertyPath());
        }

        /// <summary>
        /// Includes the specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        public ILazyLoaderWithInclude<T> Include(Expression<Func<T, IEnumerable<string>>> path)
        {
            return Include(path.ToPropertyPath());
        }

        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <param name="ids">The ids.</param>
        public Lazy<Dictionary<string, T>> Load(params string[] ids)
        {
            return session.LazyLoadInternal<T>(ids, includes.ToArray(), null);
        }

        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <param name="ids">The ids.</param>
        /// <returns></returns>
        public Lazy<Dictionary<string, T>> Load(IEnumerable<string> ids)
        {
            return session.LazyLoadInternal<T>(ids.ToArray(), includes.ToArray(), null);
        }

        /// <summary>
        /// Loads the specified id.
        /// </summary>
        public Lazy<T> Load(string id)
        {
            var results = session.LazyLoadInternal<T>(new[] { id }, includes.ToArray(), null);
            return new Lazy<T>(() => results.Value.Values.First());
        }
        
        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <typeparam name="TResult"></typeparam>
        /// <param name="ids">The ids.</param>
        public Lazy<Dictionary<string, TResult>> Load<TResult>(params string[] ids)
        {
            return session.LazyLoadInternal<TResult>(ids, includes.ToArray(), null);
        }

        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <param name="ids">The ids.</param>
        /// <returns></returns>
        public Lazy<Dictionary<string, TResult>> Load<TResult>(IEnumerable<string> ids)
        {
            return session.LazyLoadInternal<TResult>(ids.ToArray(), includes.ToArray(), null);
        }

        /// <summary>
        /// Loads the specified id.
        /// </summary>
        /// <typeparam name="TResult"></typeparam>
        /// <param name="id">The id.</param>
        public Lazy<TResult> Load<TResult>(string id)
        {
            var lazy = Load<TResult>(new[] { id });
            return new Lazy<TResult>(() => lazy.Value.Values.FirstOrDefault());
        }
    }

    public class AsyncLazyMultiLoaderWithInclude<T> : IAsyncLazyLoaderWithInclude<T>
    {
        private readonly IAsyncDocumentSessionImpl session;
        private readonly List<string> includes = new List<string>();

        /// <summary>
        /// Initializes a new instance of the <see cref="LazyMultiLoaderWithInclude{T}"/> class.
        /// </summary>
        /// <param name="session">The session.</param>
        internal AsyncLazyMultiLoaderWithInclude(IAsyncDocumentSessionImpl session)
        {
            this.session = session;
        }

        /// <summary>
        /// Includes the specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        public  IAsyncLazyLoaderWithInclude<T>  Include(string path)
        {
            includes.Add(path);
            return this;
        }

        /// <summary>
        /// Includes the specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        public IAsyncLazyLoaderWithInclude<T> Include(Expression<Func<T, string>> path)
        {
            return Include(path.ToPropertyPath());
        }


        /// <summary>
        /// Includes the specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        public IAsyncLazyLoaderWithInclude<T> Include(Expression<Func<T, IEnumerable<string>>> path)
        {
            return Include(path.ToPropertyPath());
        }

        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <param name="ids">The ids.</param>
        public Lazy<Task<Dictionary<string, T>>> LoadAsync(params string[] ids)
        {
            return session.LazyAsyncLoadInternal<T>(ids, includes.ToArray(), null);
        }

        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <param name="ids">The ids.</param>
        /// <returns></returns>
        public Lazy<Task<Dictionary<string, T>>> LoadAsync(IEnumerable<string> ids)
        {
            return session.LazyAsyncLoadInternal<T>(ids.ToArray(), includes.ToArray(), null);
        }

        /// <summary>
        /// Loads the specified id.
        /// </summary>
        public Lazy<Task<T>> LoadAsync(string id)
        {
            var results = session.LazyAsyncLoadInternal<T>(new[] { id }, includes.ToArray(), null);

            return new Lazy<Task<T>>(() => results.Value.ContinueWith(x => x.Result.Values.First()));
        }

        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <typeparam name="TResult"></typeparam>
        /// <param name="ids">The ids.</param>
        public Lazy<Task<Dictionary<string, TResult>>> LoadAsync<TResult>(params string[] ids)
        {
            return session.LazyAsyncLoadInternal<TResult>(ids, includes.ToArray(), null); 
        }

        /// <summary>
        /// Loads the specified ids.
        /// </summary>
        /// <param name="ids">The ids.</param>
        /// <returns></returns>
        public Lazy<Task<Dictionary<string, TResult>>> LoadAsync<TResult>(IEnumerable<string> ids)
        {
            return session.LazyAsyncLoadInternal<TResult>(ids.ToArray(), includes.ToArray(), null);
        }

        /// <summary>
        /// Loads the specified id.
        /// </summary>
        /// <typeparam name="TResult"></typeparam>
        /// <param name="id">The id.</param>
        public Lazy<Task<TResult>> LoadAsync<TResult>(string id)
        {
            var lazy = LoadAsync<TResult>(new[] { id });
            return new Lazy<Task<TResult>>(() => lazy.Value.ContinueWith(x => x.Result.Values.First()));
        }
    }
}

